### PR TITLE
Add HTTP peer connection suspect state

### PR DESCRIPTION
**Pull request by kriskowal**
This change causes the HTTP transport to suspect that a peer might be
disconnected if a request times out.
In some cases, this indicates that the connection is half open, that the
destination died without sending a TCP FIN packet.
In the suspect state, the peer will remain available while the peer manager
attempts to make a new connection in the background.
If the connection manager fails, the peer will be demoted to unavailable.

- [x] Ensure that the connection management loop backs off when available but 
  suspect.

### CODE REVIEW

**Comment by AlexAPB** [comment]
💯  seems like good way to start incorporating health checks into peer management

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,9 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
+- No changes yet.
```
Comment (abhinav) [inline CHANGELOG.md]: Did you mean to include this change in this PR?

File: transport/http/peer.go [hunk 2]
```diff
@@ -98,7 +129,6 @@ func (p *httpPeer) MaintainConn() {
 	// Attempt to retain an open connection to each peer so long as it is
 	// retained.
 	for {
-		p.Peer.SetStatus(peer.Connecting)
```
Comment (prashantv) [inline transport/http/peer.go]: is it safe to remove this? seems like we're losing track of this state, which might be useful for debugging.

File: transport/tchannel/header.go [hunk 3]
```diff
@@ -40,10 +40,10 @@ const (
 	ErrorMessageHeaderKey = "$rpc$-error-message"
 )
 
-var _reservedHeaderKeys = map[string]bool{
-	ErrorCodeHeaderKey:    true,
-	ErrorNameHeaderKey:    true,
-	ErrorMessageHeaderKey: true,
+var _reservedHeaderKeys = map[string]struct{}{
```
Comment (prashantv) [inline transport/tchannel/header.go]: unrelated changes?

File: transport/http/peer.go [hunk 4]
```diff
@@ -98,7 +129,6 @@ func (p *httpPeer) MaintainConn() {
...
 		if p.isAvailable() {
 			p.Peer.SetStatus(peer.Available)
```
[the remark this answers is not in the document]
Comment (kriskowal) [reply transport/http/peer.go]: It’s technically fine since Connecting isn’t particularly useful, but I’ve found a way to keep it, by moving it to OnDisconnect and after exponential back-off, to ensure that the state remains steadily Available if we re-enter the loop due to OnSuspect.

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,18 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ## [Unreleased]
 ### Changed
 - Enabled random shuffling of peerlist order by default.
+- The HTTP protocol now mitigates peers that are unavailable due to a half-open
```
[the remark this answers is not in the document]
Comment (kriskowal) [reply CHANGELOG.md]: I’m inclined to keep it so upgraders can reason about the behavior if they observe an unexpected change.

File: transport/tchannel/header.go (hunk 3 above)
Comment (kriskowal) [reply transport/tchannel/header.go]: Shucks, you noticed.

File: internal/examples/thrift-hello/hello/main.go [hunk 6]
```diff
@@ -31,14 +31,27 @@ import (
...
 )
 
+var count int
+
 var flagWait = flag.Bool("wait", false, "Wait for a signal to exit")
 
+var flagRR = flag.Bool("round-robin", false, "Use round-robin instead of fewest-pending-requests load balancer")
+
+func init() {
+	flag.IntVar(&count, "count", 1, "How many requests to send, or -1 to run indefinitely")
```
Comment (prashantv) [inline internal/examples/thrift-hello/hello/main.go]: can we group all the flags together,
```
var (
  flagWait = ...
  flagRR = ...
  flagInt = ...
)
```

File: internal/examples/thrift-hello/hello/main.go [hunk 7]
```diff
@@ -78,16 +108,21 @@ func do() error {
 	// using the dispatcher and associated "hello" outbound
 	client := helloclient.New(dispatcher.ClientConfig("hello"))
 
-	// build a context with a 1 second deadline
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
-	defer cancel()
+	for i := 0; i != count; i++ {
```
Comment (prashantv) [inline internal/examples/thrift-hello/hello/main.go]: can we just use `flag.Int` and use `*count` here? doesn't seem worth it to declare flags in different ways

File: transport/http/constants.go [hunk 8]
```diff
@@ -26,6 +26,8 @@ const transportName = "http"
 
 var defaultConnTimeout = 500 * time.Millisecond
 
+var defaultInnocenceWindow = 5 * time.Second
```
Comment (prashantv) [inline transport/http/constants.go]: nit: could use var group here

File: transport/http/peer.go [hunk 9]
```diff
@@ -73,8 +80,32 @@ func (p *httpPeer) isAvailable() bool {
...
 }
 
+func (p *httpPeer) OnSuspect() {
+	now := time.Now().UnixNano()
+	innocentUntil := p.innocentUntilUnixNano.Load()
+
+	// Do not check for connectivity after every request timeout.
+	// Spread them out so they only occur once in every innocence window.
+	if now < innocentUntil {
+		return
```
Comment (prashantv) [inline transport/http/peer.go]: no coverage here

**Comment by peter-edge** [comment]
What's the status on this?

File: transport/http/peer.go (hunk 9 above)
Comment (kriskowal) [reply transport/http/peer.go]: Addressed

**Comment by kriskowal** [comment]
Status: pending @prashantv’s acceptance.

I’ve rebased and improve the test coverage somewhat. Producing the circumstances that this feature mitigates is probably not worth the cost in a unit test. We validated this change by manipulating iptables in a virtual Linux host, using the Thrift example and yab.

File: transport/http/peer.go (hunk 9 above)
Comment (prashantv) [reply transport/http/peer.go]: still red according to the coverage, but it could be stale?

File: transport/http/peer_test.go [hunk 10]
```diff
@@ -102,6 +104,43 @@ func TestHTTPWithRoundRobin(t *testing.T) {
...
+	go server.Stop()
+	ctx = context.Background()
+	ctx, cancel4 := context.WithTimeout(ctx, 5*testtime.Millisecond)
+	defer cancel4()
+	for {
+		err := integrationtest.Call(ctx, c)
+		if err != nil {
+			break
+		}
+	}
```
Comment (prashantv) [inline transport/http/peer_test.go]: can we check that the peer is no longer available here? e.g., how do we know whether this call is failing because the server is stopped but the client still tries to send it requests, vs call is failing because there are no peer available? perhaps we just need to test the error

File: transport/http/peer_test.go [hunk 11]
```diff
@@ -102,6 +104,43 @@ func TestHTTPWithRoundRobin(t *testing.T) {
...
+func TestHTTPOnSuspect(t *testing.T) {
+	server, serverAddr := spec.NewServer(t, ":0")
+
+	client, c := spec.NewClient(t, []string{serverAddr})
+	defer client.Stop()
+
+	// Exercise OnSuspect
+	ctx := context.Background()
+	ctx, cancel1 := context.WithTimeout(ctx, 5*testtime.Millisecond)
+	defer cancel1()
```
Comment (prashantv) [inline transport/http/peer_test.go]: do we need a new `cancel` name? should we just do `ctx, cancel = context.WithTimeout(...)`?

File: transport/http/peer.go (hunk 9 above)
Comment (kriskowal) [reply transport/http/peer.go]: I added a runtime.Gosched() to the innocence window test and it gets coverage half the time now. I think that ought to be enough, unless there’s something not-too-invasive I can do to force the test and OnSuspect() call to synchronize. For now, I’m going to land this.
